Plugins/Sycl/src/Seeding/CreateSeedsForGroupSycl.cpp

```diff
@@ -113,77 +99,128 @@ void createSeedsForGroupSycl(
+                                                  std::vector<std::size_t>(M,B),
+                                                  *device_resource, &resource);
+      midTopDupletBuffer = std::make_unique<vecmem::data::jagged_vector_buffer<uint32_t>>
+                                                (std::vector<std::size_t>(M, 0),
+                                                  std::vector<std::size_t>(M, T),
+                                                  *device_resource, &resource);
+    }
+    copy.setup(*midBotDupletBuffer);
+    copy.setup(*midTopDupletBuffer);
+
     // Atomic accessor type used throughout the code.
     using AtomicAccessor =
         sycl::ONEAPI::atomic_accessor<uint32_t, 1,
                                       sycl::ONEAPI::memory_order::relaxed,
                                       sycl::ONEAPI::memory_scope::device>;
-    {
-      // Allocate temporary buffers on top of the count arrays booked in USM.
-      sycl::buffer<uint32_t> countBotBuf(countBotDuplets.data(), M);
-      sycl::buffer<uint32_t> countTopBuf(countTopDuplets.data(), M);
-
-      // Perform the middle-bottom duplet search.
-      q->submit([&](cl::sycl::handler& h) {
-        AtomicAccessor countBotDupletsAcc(countBotBuf, h);
-        detail::DupletSearch<detail::SpacePointType::Bottom, AtomicAccessor>
-            kernel(M, deviceMiddleSPs, B, deviceBottomSPs, deviceTmpIndBot,
-                   countBotDupletsAcc, seedfinderConfig);
-        h.parallel_for<class DupletSearchBottomKernel>(bottomDupletNDRange,
-                                                       kernel);
-      });
-
-      // Perform the middle-top duplet search.
-      q->submit([&](cl::sycl::handler& h) {
-        AtomicAccessor countTopDupletsAcc(countTopBuf, h);
-        detail::DupletSearch<detail::SpacePointType::Top, AtomicAccessor>
-            kernel(M, deviceMiddleSPs, T, deviceTopSPs, deviceTmpIndTop,
-                   countTopDupletsAcc, seedfinderConfig);
-        h.parallel_for<class DupletSearchTopKernel>(topDupletNDRange, kernel);
-      });
-    }  // sync (buffers get destroyed and duplet counts are copied back to
-       // countBotDuplets and countTopDuplets)
-
+  {
+
+    // Perform the middle-bottom duplet search.
+    q->submit([&](cl::sycl::handler& h) {
+      detail::DupletSearch<detail::SpacePointType::Bottom>
+          kernel(middleSPsView, bottomSPsView,
+                 *midBotDupletBuffer, seedfinderConfig);
+      h.parallel_for<class DupletSearchBottomKernel>(bottomDupletNDRange,
+                                                     kernel);
+    });
+
+    // Perform the middle-top duplet search.
+    q->submit([&](cl::sycl::handler& h) {
+      detail::DupletSearch<detail::SpacePointType::Top>
+          kernel(middleSPsView, topSPsView,
+                 *midTopDupletBuffer, seedfinderConfig);
+      h.parallel_for<class DupletSearchTopKernel>(topDupletNDRange, kernel);
+    });
```

Comment:
:thinking: Since no buffers are used in these kernels anymore, don't we need to explicitly use `wait_and_throw()` somewhere? Like:

```c++
auto middleBottomEvent = q->submit(...);
auto middleTopEvent = q->submit(...);
middleBottomEvent.wait_and_throw();
middleTopEvent.wait_and_throw();
```

I think we do... (Note that you should still launch the two kernels "in parallel", and only wait for them finishing afterwards.)

Reply:
Done.